### PR TITLE
feat: bulk delete dashboard api

### PR DESCRIPTION
**Pull request by tracy-french**
# Description

This change improves the client-side bulk delete experience by implementing `/dashboards/bulk` delete API.

Smaller changes:
- Remove dashboard caching - We may not have a single server and the caching will not work properly if there are more than one.
- Revise `either` naming in dashboard controller to `result`.
- Update `core` code coverage requirements.
- Prevent code coverage from running on single commit tests - It incorrectly runs the coverage on the single commit and is easy to fail.

## Legal

This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: apps/core/src/dashboards/dashboards.service.ts [hunk 1]
```diff
@@ -29,4 +29,9 @@ export class DashboardsService {
   public async delete(id: Dashboard['id']) {
     return this.repository.delete(id);
   }
+
+  public async bulkDelete(ids: Dashboard['id'][]) {
+    // TODO: use BatchWriteItem instead of Promise.all
```
Comment (dpportet) [inline apps/core/src/dashboards/dashboards.service.ts:34]: Is there anything preventing us from using BatchWriteItem now instead of later?

File: apps/core/src/dashboards/dashboards.e2e.spec.ts [hunk 2]
```diff
@@ -1284,6 +1284,116 @@ describe('DashboardsModule', () => {
...
+        method: 'GET',
+        url: `/dashboards/${dashboard3.id}`,
+      });
+
+      expect(getResponse1.statusCode).toBe(404);
+      expect(getResponse2.statusCode).toBe(404);
+      expect(getResponse3.statusCode).toBe(404);
+    });
+
+    test('returns 200 when called with empty id list', async () => {
```
Comment (dpportet) [inline apps/core/src/dashboards/dashboards.e2e.spec.ts:1338]: Can you add one more test for 'returns 200 when called with a single item list'? I think that can sometimes be an edge case when moving from single-item to multi-item data.

File: apps/core/src/dashboards/dashboards.controller.ts [hunk 3]
```diff
@@ -112,10 +114,41 @@ export class DashboardsController {
...
 
     return either.ok;
   }
+
+  @HttpCode(200)
+  @Delete('bulk')
+  public async bulkDelete(
+    @Body() bulkDeleteDashboardsDto: BulkDeleteDashboardDto,
+  ): Promise<{ deletedIds: Dashboard['id'][] }> {
+    // cache is invalid after deleting a dashboard
```
Comment (dpportet) [inline apps/core/src/dashboards/dashboards.controller.ts:129]: why is it invalid after deleting a dashboard?

File: apps/core/src/dashboards/dashboards.controller.ts [hunk 4]
```diff
@@ -112,10 +114,41 @@ export class DashboardsController {
...
+    await this.cacheManager.reset();
+
+    const eithers = await this.dashboardsService.bulkDelete(
```
Comment (dpportet) [inline apps/core/src/dashboards/dashboards.controller.ts:132]: What's the reasoning behind calling this `eithers`? It is kinda hard to read for me, I would expect it to be something like `response`.

**Review by dpportet — CHANGES_REQUESTED** [review]

File: apps/core/src/dashboards/dashboards.service.ts (hunk 1 above)
Comment (tracy-french) [reply apps/core/src/dashboards/dashboards.service.ts:34]: Not really. It's more work. I'm pro-iterative development. We can do this change in a follow up.

File: apps/core/src/dashboards/dashboards.e2e.spec.ts (hunk 2 above)
Comment (tracy-french) [reply apps/core/src/dashboards/dashboards.e2e.spec.ts:1338]: Yup, will do.

File: apps/core/src/dashboards/dashboards.controller.ts (hunk 3 above)
Comment (tracy-french) [reply apps/core/src/dashboards/dashboards.controller.ts:129]: The cache could contain dashboards which have been deleted. We currently aren't working with the cache in a granular manner and are just nuking it on changes. The better method would be to mutate the cache to remove the deleted dashboards. We're probably gonna remove the cache though, as it won't work if we have multiple application instances, so not something for us to deal with now.

File: apps/core/src/dashboards/dashboards.controller.ts (hunk 4 above)
Comment (tracy-french) [reply apps/core/src/dashboards/dashboards.controller.ts:132]: Yeah, this is a mistake. It should be `result`. Would that be a better name?

File: apps/core/src/dashboards/dashboards.service.ts (hunk 1 above)
Comment (chejimmy) [reply apps/core/src/dashboards/dashboards.service.ts:34]: If we are doing batch, we will want to make sure each dashboard's summary and definition are deleted together. `TransactWriteItems` would probably be the safer API to use so the deletion is atomic. However, it has a limit of 100 items so we would want to keep that in mind ([doc](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_TransactWriteItems.html#DDB-TransactWriteItems-request-TransactItems))

Comment (chejimmy) [reply apps/core/src/dashboards/dashboards.service.ts:34]: If we are calling the batch API, we need to keep in mind of the atomicity. We want to make sure each dashboard's summary and definition are delete together. [TransactWriteItems](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_TransactWriteItems.html) might be the better API; however, keep in mind of its [limit of 100 items](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_TransactWriteItems.html#DDB-TransactWriteItems-request-TransactItems) and constraint of no partial update/deletion.

File: apps/core/src/dashboards/dashboards.service.ts [hunk 5]
```diff
@@ -29,4 +29,9 @@ export class DashboardsService {
...
+    return Promise.all(ids.map((id) => this.repository.delete(id)));
```
Comment (chejimmy) [inline apps/core/src/dashboards/dashboards.service.ts:35]: I don't think [Promise.all(...)](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all) can handle multiple rejections
> The rejection reason is the rejection reason of the first promise that was rejected.

is that the desirable behavior or [Promise.allSettled()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled)'s behavior would serve better?

File: apps/client/src/routes/dashboards/dashboards-index/components/delete-dashboard-modal.tsx [hunk 6]
```diff
@@ -59,18 +66,36 @@ export function DeleteDashboardModal(props: DeleteDashboardModalProps) {
...
+                  intl.formatMessage({
+                    defaultMessage:
+                      'An error occured, but partial deletion was successful.',
+                    description: 'partial deletion successful message',
+                  }),
+                ),
+              ),
+            );
+          }
           emit(new GenericErrorNotification(error));
```
[the remark this answers is not in the document]
Comment (tracy-french) [reply apps/client/src/routes/dashboards/dashboards-index/components/delete-dashboard-modal.tsx:95]: Oops! Good catch!

File: apps/client/src/routes/dashboards/dashboards-index/hooks/use-delete-dashboard-mutation.ts [hunk 7]
```diff
@@ -1,12 +1,10 @@
...
-import type { DashboardSummary } from '~/services';
+import { bulkDeleteDashboards } from '~/services';
+import { invalidateDashboards } from '~/data/dashboards';
 
 export function useDeleteDashboardMutation() {
   return useMutation({
-    mutationFn: (dashboard: DashboardSummary) => deleteDashboard(dashboard.id),
-    onSuccess: (_data, variables) => void invalidateDashboard(variables.id),
+    mutationFn: bulkDeleteDashboards,
+    onSuccess: () => void invalidateDashboards(),
```
[the remark this answers is not in the document]
Comment (tracy-french) [reply apps/client/src/routes/dashboards/dashboards-index/hooks/use-delete-dashboard-mutation.ts:8]: I'll need to check this out. My thinking was it didn't matter, since this would cause the table page to refresh. I'll need to check if the dashboard page is still accessible. If it is, I'll need to update the logic.

[the remark this answers is not in the document]
Comment (tracy-french) [reply apps/client/src/routes/dashboards/dashboards-index/hooks/use-delete-dashboard-mutation.ts:8]: We're good. The only way to access the dashboard page after deletion is through the "create dashboard" success notification, which may still be visible when the dashboard is deleted. The single dashboard cache only hydrates when visiting the dashboard page.

The "create dashboard" success notification disappears once you leave the dashboard list page. Although the "update dashboard" function calls `GET /dashboards/{dashboardId}` during inline table editing, it doesn't hydrate the cache. This means that even if you try to access the dashboard page via the success notification after deleting it, you'll be directed to a 404 error page.

File: apps/core/src/dashboards/dashboards.controller.ts (hunk 4 above)
Comment (dpportet) [reply apps/core/src/dashboards/dashboards.controller.ts]: Yes, thanks.

**Review by dpportet — APPROVED** [review]

**Review by chejimmy — APPROVED** [review]
LGTM 🚀

**Review by chejimmy — APPROVED** [review]
LGTM 🚀🚀🚀
